src/lib.rs

```diff
@@ -1195,6 +1195,66 @@ fn graph_adjacency_matrix(
     Ok(matrix.into_pyarray(py).into())
 }
 
+#[pyfunction]
+fn graph_all_simple_paths(
+    graph: &graph::PyGraph,
+    from: usize,
+    to: usize,
+    min_depth: Option<usize>,
+    cutoff: Option<usize>,
+) -> PyResult<Vec<Vec<usize>>> {
+    let from_index = NodeIndex::new(from);
+    let to_index = NodeIndex::new(to);
+    let min_intermediate_nodes: usize = match min_depth {
+        Some(depth) => depth - 2,
+        None => 0,
+    };
+    let cutoff_petgraph: Option<usize> = match cutoff {
+        Some(depth) => Some(depth - 2),
+        None => None,
+    };
+    let result: Vec<Vec<usize>> = algo::all_simple_paths(
+        graph,
+        from_index,
+        to_index,
+        min_intermediate_nodes,
+        cutoff_petgraph,
+    )
+    .map(|v: Vec<NodeIndex>| v.into_iter().map(|i| i.index()).collect())
+    .collect();
+    Ok(result)
+}
+
+#[pyfunction]
+fn dag_all_simple_paths(
```

Comment:
It's not possible for one function to implement the feature for both PyDAG and PyGraph? It looks like these functions are the same.

Reply:
I think it should be possible, but I haven't found a good pattern for doing this yet. The way it would have to be done is defining a trait (or a set of traits) which are defined for both `PyDAG` and `PyGraph` that cover all the required methods and attributes that these functions required. But I've not been able to find a way to do this yet. It would be something like what is done here: https://github.com/Qiskit/retworkx/blob/master/src/dag_isomorphism.rs#L263 Until I figure that out, I've just been duplicating all the functions over both types. I figured when I do come up with a pattern that works I can just deprecate the class specific functions and replace these with `all_simple_paths()` and replace both of these implementations with a call to that.

Reply:
I tried to play with this a while today, and haven't made much progress. There seems to be a conflict between all the input traits required from petgraph's all_simple_paths and the pyfunction macro that pyo3 uses. The closest I've found is to have a function signature like:

```rust
#[pyfunction]
fn all_simple_paths<'a, T>(
    graph: T,
    from: usize,
    to: usize,
    min_depth: Option<usize>,
    cutoff: Option<usize>,
) -> PyResult<Vec<Vec<usize>>>
where
    T: FromPyObject<'a>
        + petgraph::visit::IntoNeighborsDirected<
            EdgeId = EdgeIndex,
            NodeId = NodeIndex,
            Neighbors = petgraph::stable_graph::Neighbors<'a, PyObject>,
            NeighborsDirected = petgraph::stable_graph::Neighbors<'a, PyObject>,
        > + GraphBase<NodeId = NodeIndex, EdgeId = EdgeIndex>
        + petgraph::visit::NodeCount,
    T::NodeId: Eq + std::hash::Hash,
{
```

but then rust complains about the bounds traits not being implemented for the input `()`. 

```
error[E0277]: the trait bound `(): petgraph::visit::IntoNeighborsDirected` is not satisfied
    --> src/lib.rs:1199:4
     |
1199 |   fn all_simple_paths<'a, T>(
     |      ^^^^^^^^^^^^^^^^
     |      |
     |      the trait `petgraph::visit::IntoNeighborsDirected` is not implemented for `()`
     |      required by a bound in this
...
1208 |           + petgraph::visit::IntoNeighborsDirected<
     |  ___________-
1209 | |             EdgeId = EdgeIndex,
1210 | |             NodeId = NodeIndex,
1211 | |             Neighbors = petgraph::stable_graph::Neighbors<'a, PyObject>,
1212 | |             NeighborsDirected = petgraph::stable_graph::Neighbors<'a, PyObject>,
1213 | |         > + GraphBase<NodeId = NodeIndex, EdgeId = EdgeIndex>
     | |_________- required by this bound in `all_simple_paths`

error[E0277]: the trait bound `(): petgraph::visit::NodeCount` is not satisfied
    --> src/lib.rs:1199:4
     |
1199 | fn all_simple_paths<'a, T>(
     |    ^^^^^^^^^^^^^^^^
     |    |
     |    the trait `petgraph::visit::NodeCount` is not implemented for `()`
     |    required by a bound in this
...
1214 |         + petgraph::visit::NodeCount,
     |           -------------------------- required by this bound in `all_simple_paths`

error[E0277]: the trait bound `(): pyo3::PyClass` is not satisfied
    --> src/lib.rs:1199:4
     |
1199 | fn all_simple_paths<'a, T>(
     |    ^^^^^^^^^^^^^^^^
     |    |
     |    the trait `pyo3::PyClass` is not implemented for `()`
     |    required by a bound in this
...
1207 |     T: FromPyObject<'a>
     |        ---------------- required by this bound in `all_simple_paths`
     |
     = note: required because of the requirements on the impl of `pyo3::FromPyObject<'_>` for `()`

error[E0277]: the trait bound `(): pyo3::PyClass` is not satisfied
    --> src/lib.rs:1198:1
     |
1198 | #[pyfunction]
     | ^^^^^^^^^^^^^ the trait `pyo3::PyClass` is not implemented for `()`
     |
     = note: required because of the requirements on the impl of `pyo3::FromPyObject<'_>` for `()`
     = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 4 previous errors
```

I'll probably  have to reach out for help from the pyo3 maintainers to figure out how to do this or if it's even possible. I'm probably just overlooking something simple, but it might be a while before I can come up with an answer for this. That being said I think just living with 2 funcs for now doesn't seem like a big deal until we can figure this out.

Reply:
yeah, that seems fine. thanks for looking into it and explaining :))